File: 16-07-2024/Highest-Lowest-Frequency-Element/Solutions/Solution.py
def maxFrequency(self, nums, k):
    l, h, ans = 1, len(nums), 0
    nums.sort()

    while l <= h:
        mid = l + (h - l) // 2
        if self.pos(nums, mid, k):
            ans = mid
            l = mid + 1
        else:
            h = mid - 1

    return ans

def pos(self, nums, mid, k):
    wsum = 0
    for i in range(mid):
        wsum += nums[i]

    tsum = nums[mid - 1] * mid

    if tsum - wsum <= k:
        return True

    j = 0
    for i in range(mid, len(nums)):
        wsum -= nums[j]
        wsum += nums[i]
        tsum = nums[i] * mid

        if tsum - wsum <= k:
            return True
        j += 1

    return False

File: 16-07-2024/Highest-Lowest-Frequency-Element/Solutions/Solution.cpp
int maxFrequency(vector<int>& nums, int k) {
    int l = 1, h = nums.size(), ans = 0;
    sort(nums.begin(), nums.end());

    while (l <= h) {
        int mid = l + (h - l) / 2;
        if (pos(nums, mid, k)) {
            ans = mid;
            l = mid + 1;
        } else {
            h = mid - 1;
        }
    }

    return ans;
}

bool pos(vector<int>& nums, int mid, int k) {
    long long wsum = 0, tsum = 0;
    for (int i = 0; i < mid; ++i)
        wsum += nums[i];

    tsum = static_cast<long long>(nums[mid - 1]) * mid;

    if (tsum - wsum <= k)
        return true;

    int j = 0;
    for (int i = mid; i < nums.size(); ++i) {
        wsum -= nums[j];
        wsum += nums[i];
        tsum = static_cast<long long>(nums[i]) * mid;

        if (tsum - wsum <= k)
            return true;
        j++;
    }

    return false;
}